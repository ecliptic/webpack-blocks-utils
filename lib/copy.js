'use strict'

exports.__esModule = true
exports.copyOptions = exports.copyPattern = exports.copy = undefined

var _extends =
  Object.assign ||
  function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i]
      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key]
        }
      }
    }
    return target
  }
/**
 * Webpack block for copy-webpack-plugin
 *
 * @see https://github.com/kevlened/copy-webpack-plugin
 */

var _copyWebpackPlugin = require('copy-webpack-plugin')

var _copyWebpackPlugin2 = _interopRequireDefault(_copyWebpackPlugin)

function _interopRequireDefault (obj) {
  return obj && obj.__esModule ? obj : {default: obj}
}

/**
 * Adds a simple copy pattern to the list of patterns for the plugin.
 */
function copy (from, to) {
  return Object.assign(
    function (context) {
      return function (prevConfig) {
        context.copyPlugin = context.copyPlugin || {patterns: []}

        // Merge the provided simple pattern into the config
        context.copyPlugin = _extends({}, context.copyPlugin, {
          patterns: [].concat(context.copyPlugin.patterns, [
            {from: from, to: to},
          ]),
        })

        // Don't alter the config yet
        return prevConfig
      }
    },
    {post: postConfig}
  )
}

/**
 * Adds an advanced pattern to the list of patterns for the plugin.
 */
function copyPattern (pattern) {
  return Object.assign(
    function (context) {
      return function (prevConfig) {
        context.copyPlugin = context.copyPlugin || {patterns: []}

        // Merge the provided advanced pattern into the config
        context.copyPlugin = _extends({}, context.copyPlugin, {
          patterns: [].concat(context.copyPlugin.patterns, [pattern]),
        })

        // Don't alter the config yet
        return prevConfig
      }
    },
    {post: postConfig}
  )
}

/**
 * Sets options for the copy plugin.
 */
function copyOptions (options) {
  return Object.assign(
    function (context) {
      return function (prevConfig) {
        context.copyPlugin = context.copyPlugin || {}

        // Merge the provided copy plugin config into the context
        context.copyPlugin = _extends({}, context.copyPlugin, {
          options: options,
        })

        // Don't alter the config yet
        return prevConfig
      }
    },
    {post: postConfig}
  )
}

/**
 * Instantiate the copy plugin.
 */
function postConfig (context, _ref) {
  var merge = _ref.merge

  return function (prevConfig) {
    var _context$copyPlugin = context.copyPlugin,
      patterns = _context$copyPlugin.patterns,
      options = _context$copyPlugin.options

    var plugin = new _copyWebpackPlugin2.default(patterns, options)
    return merge({plugins: [plugin]})
  }
}

exports.copy = copy
exports.copyPattern = copyPattern
exports.copyOptions = copyOptions
