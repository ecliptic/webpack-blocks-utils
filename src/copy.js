// @flow
/**
 * Webpack block for copy-webpack-plugin
 *
 * @see https://github.com/kevlened/copy-webpack-plugin
 */
import CopyWebpackPlugin from 'copy-webpack-plugin'
import type {Config, Helpers, Block} from './types'

export type Pattern = {
  from: string | Object,
  to?: string,
  toType?: "file" | "dir" | "template",
  context?: string,
  flatten?: boolean,
  ignore?: Array<string>,
  transform?: (content: string, path: string) => string,
  force?: boolean
}

export type Options = {
  ignore?: Array<string>,
  copyUnmodified?: boolean,
  debug?: "warning" | "info" | "debug"
}

/**
 * Adds a simple copy pattern to the list of patterns for the plugin.
 */
function copy (from: string, to?: string): Block {
  return Object.assign(
    context => prevConfig => {
      context.copyPlugin = context.copyPlugin || {patterns: []}

      // Merge the provided simple pattern into the config
      context.copyPlugin = {
        ...context.copyPlugin,
        patterns: [...context.copyPlugin.patterns, {from, to}],
      }

      // Don't alter the config yet
      return prevConfig
    },
    {post: postConfig}
  )
}

/**
 * Adds an advanced pattern to the list of patterns for the plugin.
 */
function copyPattern (pattern: Pattern): Block {
  return Object.assign(
    context => prevConfig => {
      context.copyPlugin = context.copyPlugin || {patterns: []}

      // Merge the provided advanced pattern into the config
      context.copyPlugin = {
        ...context.copyPlugin,
        patterns: [...context.copyPlugin.patterns, pattern],
      }

      // Don't alter the config yet
      return prevConfig
    },
    {post: postConfig}
  )
}

/**
 * Sets options for the copy plugin.
 */
function copyOptions (options: Options): Block {
  return Object.assign(
    context => prevConfig => {
      context.copyPlugin = context.copyPlugin || {}

      // Merge the provided copy plugin config into the context
      context.copyPlugin = {...context.copyPlugin, options}

      // Don't alter the config yet
      return prevConfig
    },
    {post: postConfig}
  )
}

/**
 * Instantiate the copy plugin.
 */
function postConfig (context: Object, {merge}: Helpers): Config => Config {
  const {patterns, options} = context.copyPlugin
  const plugin = new CopyWebpackPlugin(patterns, options)
  return merge({plugins: [plugin]})
}

export {copy, copyPattern, copyOptions}
