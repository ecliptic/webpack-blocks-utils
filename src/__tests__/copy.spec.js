// @flow
import {copy, copyPattern, copyOptions} from '../copy'

describe('copy', () => {
  describe('copy()', () => {
    test('from only', () => {
      const block = copy('assets/robots.txt')
      const context = {}
      const helpers = {
        merge: configSnippet => {
          expect(configSnippet.plugins[0]).toHaveProperty('apply')
          return prevConfig => configSnippet
        },
      }

      const intermediateConfig = block(context, helpers)({})
      if (block.post) block.post(context, helpers)(intermediateConfig)

      expect(context).toMatchSnapshot()
    })

    test('from and to', () => {
      const block = copy('assets/robots.txt', 'public/assets/robots.txt')
      const context = {}
      const helpers = {
        merge: configSnippet => {
          expect(configSnippet.plugins[0]).toHaveProperty('apply')
          return prevConfig => configSnippet
        },
      }

      const intermediateConfig = block(context, helpers)({})
      if (block.post) block.post(context, helpers)(intermediateConfig)

      expect(context).toMatchSnapshot()
    })

    test('multiple definitions', () => {
      const context = {}
      const helpers = {
        merge: configSnippet => {
          expect(configSnippet.plugins[0]).toHaveProperty('apply')
          return prevConfig => configSnippet
        },
      }

      // The first definition
      copy('assets/not-robots.txt')(context, helpers)({})

      // The second
      const block = copy('assets/robots.txt', 'public/assets/robots.txt')
      const intermediateConfig = block(context, helpers)({})

      if (block.post) block.post(context, helpers)(intermediateConfig)

      expect(context).toMatchSnapshot()
    })
  })

  describe('copyPattern()', () => {
    test('one pattern', () => {
      const block = copyPattern({from: 'test', to: 'target'})
      const context = {}
      const helpers = {
        merge: configSnippet => {
          expect(configSnippet.plugins[0]).toHaveProperty('apply')
          return prevConfig => configSnippet
        },
      }

      const intermediateConfig = block(context, helpers)({})
      if (block.post) block.post(context, helpers)(intermediateConfig)

      expect(context).toMatchSnapshot()
    })

    test('multiple patterns', () => {
      const context = {}
      const helpers = {
        merge: configSnippet => {
          expect(configSnippet.plugins[0]).toHaveProperty('apply')
          return prevConfig => configSnippet
        },
      }

      copyPattern({from: 'first', to: 'row'})(context, helpers)({})

      const block = copyPattern({from: 'second', to: 'row'})
      const intermediateConfig = block(context, helpers)({})
      if (block.post) block.post(context, helpers)(intermediateConfig)

      expect(context).toMatchSnapshot()
    })
  })

  test('copyOptions()', () => {
    const block = copyOptions({debug: 'warning'})
    const context = {}
    const helpers = {
      merge: configSnippet => {
        expect(configSnippet.plugins[0]).toHaveProperty('apply')
        return prevConfig => configSnippet
      },
    }

    const intermediateConfig = block(context, helpers)({})
    if (block.post) block.post(context, helpers)(intermediateConfig)

    expect(context).toMatchSnapshot()
  })
})
