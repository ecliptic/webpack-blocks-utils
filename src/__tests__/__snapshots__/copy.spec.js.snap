// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copy copy() from and to 1`] = `
Object {
  "copyPlugin": Object {
    "patterns": Array [
      Object {
        "from": "assets/robots.txt",
        "to": "public/assets/robots.txt",
      },
    ],
  },
}
`;

exports[`copy copy() from only 1`] = `
Object {
  "copyPlugin": Object {
    "patterns": Array [
      Object {
        "from": "assets/robots.txt",
        "to": undefined,
      },
    ],
  },
}
`;

exports[`copy copy() multiple definitions 1`] = `
Object {
  "copyPlugin": Object {
    "patterns": Array [
      Object {
        "from": "assets/not-robots.txt",
        "to": undefined,
      },
      Object {
        "from": "assets/robots.txt",
        "to": "public/assets/robots.txt",
      },
    ],
  },
}
`;

exports[`copy copyOptions() 1`] = `
Object {
  "copyPlugin": Object {
    "options": Object {
      "debug": "warning",
    },
  },
}
`;

exports[`copy copyPattern() multiple patterns 1`] = `
Object {
  "copyPlugin": Object {
    "patterns": Array [
      Object {
        "from": "first",
        "to": "row",
      },
      Object {
        "from": "second",
        "to": "row",
      },
    ],
  },
}
`;

exports[`copy copyPattern() one pattern 1`] = `
Object {
  "copyPlugin": Object {
    "patterns": Array [
      Object {
        "from": "test",
        "to": "target",
      },
    ],
  },
}
`;
